contributors/design-proposals/key-set-as-merge-key.md

```diff
@@ -0,0 +1,338 @@
+# Key Set as Merge Key in Strategic Merge Patch
+
+## Abstract
+
+Support a key set as the merge key in Strategic Merge Patch.
+
+## Motivation
+
+*Background*: In Strategic Merge Patch, Merge Key is the key to distinguish the entries in the list of non-primitive types.
+It must always be present and unique to perform the merge on the list of non-primitive types,
+and will be preserved.
+
+The current implementation requires a single field that uniquely identifies each element in a list.
+For some element Kinds, the identity is defined using multiple fields.
+An [example](https://github.com/kubernetes/kubernetes/issues/39188 of this is the service.spec.ports, which is identified by `protocol` and `port`.
+We need to support a key set as Merge Key. A key set must contain at least 1 key.
+
+## Proposed Change
+
+### API Change
+
+We will update the struct tags for the API resources whose Merge Key is not guaranteed to be always
+present and unique. The keys will be seperated by ",", i.e. `patchMergeKey:"<key1>,<key2>,<key3>"`.
+
+E.g. `Ports` in `ServiceSpec`.
+```go
+type ServiceSpec struct {
+  // Change patchMergeKey from "port" to "name|port"
+  Ports []ServicePort `json:"ports,omitempty" patchStrategy:"merge" patchMergeKey:"name,port" protobuf:"bytes,1,rep,name=ports"`
+  ...
+}
+```
+
+All the impacted APIs are listed in section [Impacted APIs](#impacted-apis)
+
+### Open API
+
+Update Open API schema to have a string that contains multiple keys for `patchMergeKey`.
```

Comment:
What does this mean, that the type of `patchMergeKey` needs to change? Can you give an example of the OpenAPI change?